cpp/tests/prims/mg_transform_reduce_e.cu

```diff
@@ -0,0 +1,445 @@
+#include <raft/comms/comms.hpp>
+#include <raft/comms/mpi_comms.hpp>
+#include <raft/handle.hpp>
+#include <rmm/device_scalar.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <gtest/gtest.h>
+
+#include <random>
+
+template <typename vertex_t, typename... Args>
+struct property_transform : public thrust::unary_function<vertex_t, thrust::tuple<Args...>> {
+  int mod{};
+  property_transform(int mod_count) : mod(mod_count) {}
+  constexpr __device__ auto operator()(const vertex_t& val)
+  {
+    cuco::detail::MurmurHash3_32<vertex_t> hash_func{};
+    auto value = hash_func(val) % mod;
+    return thrust::make_tuple(static_cast<Args>(value)...);
+  }
+};
+template <typename vertex_t, template <typename...> typename Tuple, typename... Args>
+struct property_transform<vertex_t, Tuple<Args...>> : public property_transform<vertex_t, Args...> {
+};
+
+template <typename Tuple, std::size_t... I>
+auto make_iterator_tuple(Tuple& data, std::index_sequence<I...>)
+{
+  return thrust::make_tuple((std::get<I>(data).begin())...);
+}
+
+template <typename... Args>
+auto get_zip_iterator(std::tuple<Args...>& data)
+{
+  return thrust::make_zip_iterator(make_iterator_tuple(
+    data, std::make_index_sequence<std::tuple_size<std::tuple<Args...>>::value>()));
+}
+
+template <typename T>
+auto get_property_iterator(std::tuple<T>& data)
+{
+  return (std::get<0>(data)).begin();
+}
+
+template <typename T0, typename... Args>
+auto get_property_iterator(std::tuple<T0, Args...>& data)
+{
+  return get_zip_iterator(data);
+}
+
+template <typename... Args>
+struct generate_impl {
+  static thrust::tuple<Args...> initial_value(int init)
+  {
+    return thrust::make_tuple(static_cast<Args>(init)...);
+  }
+  template <typename label_t>
+  static std::tuple<rmm::device_uvector<Args>...> vertex_property(rmm::device_uvector<label_t>& labels,
+                                                           int hash_bin_count,
+                                                           raft::handle_t const& handle)
+  {
+    auto data = std::make_tuple(rmm::device_uvector<Args>(labels.size(), handle.get_stream())...);
+    auto zip  = get_zip_iterator(data);
+    thrust::transform(handle.get_thrust_policy(),
+                      labels.begin(),
+                      labels.end(),
+                      zip,
+                      property_transform<label_t, Args...>(hash_bin_count));
+    return data;
+  }
+  template <typename label_t>
+  static std::tuple<rmm::device_uvector<Args>...> vertex_property(thrust::counting_iterator<label_t> begin,
+                                                           thrust::counting_iterator<label_t> end,
+                                                           int hash_bin_count,
+                                                           raft::handle_t const& handle)
+  {
+    auto length = thrust::distance(begin, end);
+    auto data   = std::make_tuple(rmm::device_uvector<Args>(length, handle.get_stream())...);
```

Comment:
This pretty much overlaps with https://github.com/rapidsai/cugraph/blob/branch-21.10/cpp/include/cugraph/utilities/dataframe_buffer.cuh#L112 and https://github.com/rapidsai/cugraph/blob/branch-21.10/cpp/include/cugraph/utilities/dataframe_buffer.cuh#L118

If you using allocate_dataframe_buffer, I think you can merge the two different versions of vertex_property to one.